scipy/optimize/_direct_py.py

```diff
@@ -0,0 +1,174 @@
+Joerg.M.Gablonsky, DIRECT Version 2.0.4. More information on the DIRECT
+algorithm can be found in Gablonsky's `thesis <http://repository.lib.ncsu.edu/ir/bitstream/1840.16/3920/1/etd.pdf>`_.
+
+.. codeauthor:: Andreas Mayer <andimscience@gmail.com>, Amit Aides <amitibo@tx.technion.ac.il>
+"""
+
+from __future__ import print_function
+import numpy as np
+from ._direct import direct
+from .optimize import OptimizeResult
+
+ERROR_MESSAGES = (
+    'u[i] < l[i] for some i',
+    'maxfun is too large',
+    'Initialization failed',
+    'There was an error in the creation of the sample points',
+    'An error occured while the function was sampled',
+    'Maximum number of levels has been reached.',
+)
+
+SUCCESS_MESSAGES = (
+    'Number of function evaluations done is larger then maxfun',
+    'Number of iterations is equal to maxiter',
+    'The best function value found is within fglper of the (known) global optimum',
+    'The volume of the hyperrectangle with best function value found is below volper',
+    'The volume of the hyperrectangle with best function value found is smaller then volper'
+)
+
+def _minimize_direct(func, bounds=None, nvar=None, args=(), disp=False,
+                     eps=1e-4, maxfun=20000, maxiter=6000, method=0,
+                     fglobal=-1e100, fglper=0.01, volper=-1.0, sigmaper=-1.0):
+    r"""
+
+    Solve an optimization problem using the DIRECT (Dividing Rectangles) algorithm.
+    It can be used to solve general nonlinear programming problems of the form:
+
+    .. math::
+
+           \min_ {x \in R^n} f(x)
+
+    subject to
+
+    .. math::
+
+           x_L \leq  x  \leq x_U
+    
+    Where :math:`x` are the optimization variables (with upper and lower
+    bounds), :math:`f(x)` is the objective function.
+
+    Parameters
+    ----------
+    func : objective function
+        called as `func(x, *args)`; does not need to be defined everywhere,
+        raise an Exception where function is not defined
+    bounds : array_like, optional
+            ``(min, max)`` pairs for each element in ``x``, defining
+            the bounds on that parameter.
+    nvar : int, optional
+        Dimensionality of x (only needed if `bounds` is not defined)
+    eps : float, optional
+        Ensures sufficient decrease in function value when a new potentially
+        optimal interval is chosen.
+    maxfun : int, optional
+        Approximate upper bound on objective function evaluations.
+        Maximal allowed value is 90000 see documentation of Fortran library.
+    maxiter : int, optional
+        Maximum number of iterations.
+        Maximal allowed value is 6000 see documentation of Fortran library.
+    method : integer, optional
+        Whether to use the original or modified DIRECT algorithm. Possible values:
+        
+        * ``method=0`` - use the original DIRECT algorithm
+        * ``method=1`` - use the modified DIRECT-l algorithm
+    fglobal : float, optional
```

Comment:
What is this parameter used for?

Reply:
This parameter is required by the algorithm. Though if it is not known then by default a very large negative value can be passed. I am having a feeling that MATLAB has a better interface for this. They accept two arguments, one a `bool` which if `True` means that global optimum value is known and if `False` then global optimum value is not known. Hence, in the latter case this parameter is ignored.

Reply:
I think this parameter is useful for testing purposes: if there is a known global minimum, by setting way we can check how many function evaluations are needed to reach it if DIRECT can solve the optimization problem.

Suggestion: set it to the minimal floating point value by default so that it is basically deactivated. Otherwise, objectives around -1e100 will be considered minimized even if all other termination criteria are not fulfilled.

Reply:
Yeah. Anyways, it is same as accepting two arguments (which is done by MATLAB code, link is in one of the comments).

Reply:
> set it to the minimal floating point value by default so that it is basically deactivated. Otherwise, objectives around -1e100 will be considered minimized even if all other termination criteria are not fulfilled.

Yes, this is correct. Ideal situation would be to go for `-float('inf')`  but I believe that is not possible with Fortran. So, smallest negative number that is possible within Fortran can be used here.